### PR TITLE
Resolve backend in rcParams.__getitem__("backend").

### PR DESCRIPTION
**Pull request by anntzer**
... to avoid leaking the non-string sentinel.

Changes to `backends/__init__` and `_backend_selection()` were necessary
to avoid import-time cycles.

The fallback order in `backend_fallback` slightly changed in that the
FooCairo backends now fallback to WxAgg instead of Wx when a wx event
loop is active, but I wouldn't worry too much about it anyways given
that the Wx backend is deprecated and the correct fallback would be
WxCairo to start with.

attn @tacaswell, xref #11844.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

**Comment by anntzer** [comment]
Test failure seems spurious, and I can't repro it locally.

**Comment by anntzer** [comment]
@tacaswell Should be fixed now.
TLDR: now `pyplot` had already been imported when pytest_configure is being called (due to the resolution system), so that call to use() needed `force=True`.

**Comment by anntzer** [comment]
... or perhaps not completely fixed :/

**Comment by anntzer** [comment]
Well, now I'm stuck with not being able to repro the failure locally (again) and rcParams["backend"] not matching the actual backend in the CI (per print-debugging), as feared in the API discussion of the original PR :/

**Comment by tacaswell** [comment]
Let see if that fixes it.  I think that some of these fixes are overlapping, but defense in depth seems like a good call here.

Turns out we were keeping state in three places we were keeping the backend state around (`rcParams`, `rcParamsDefault`, and `rcParamsOrig`).

This passes locally.

**Comment by tacaswell** [comment]
Getting closer to the problem, it looks like the rcparam tests are re-setting back to the auto sentinal which is then resolving to qt5agg and never getting reset back to 'agg'.

**Comment by tacaswell** [comment]
ok, the inputhook issues are either an IPython master-branch or pt 2 issue (down-grading both fixed it).

**Comment by anntzer** [comment]
The whole thing turned out to be more complicated than expected (hehe), but the additional fixes by @tacaswell look good.  Of course that should need at least one external reviewer...

**Comment by tacaswell** [comment]
The complexity we hit is what @timhoffm (correctly) predicted would be a problem, however I think we have no choice but to absorb this complexity as we can not break compatibility with older versions of IPython.

**Comment by anntzer** [comment]
Now that I think of it I *think* the previous approach (resolve when *writing* to the rcParams, not when reading from it) was still simpler (even if using the sentinel object rather than a list, because that object essentially never exists in the rcParams, this being the source of the issues).  After all the tests did pass on the old PR IIRC.  Of course there were the other issue pointed out by @timhoffm re: whether "active" rcParams would be a good API.

In any case things appear to work as they are (thanks for figuring it out).

File: lib/matplotlib/__init__.py [hunk 1]
```diff
@@ -1218,8 +1228,9 @@ def rc_file_defaults():
     with warnings.catch_warnings():
         warnings.simplefilter("ignore", mplDeprecation)
         from .style.core import STYLE_BLACKLIST
-        rcParams.update({k: v for k, v in rcParamsOrig.items()
-                         if k not in STYLE_BLACKLIST})
+        rcParams.update({k: rcParamsOrig[k] for k in rcParamsOrig
+                         if k not in STYLE_BLACKLIST and k != 'backend'})
+        rcParams['backend'] = dict.__getitem__(rcParamsOrig, 'backend')
```
Comment (efiring) [inline lib/matplotlib/__init__.py]: I think this is contrary to what the docstring says; if so, the docstring needs an update, and either there or in a comment it would be good to say why this exception is being made.

File: lib/matplotlib/testing/__init__.py [hunk 2]
```diff
@@ -36,7 +36,7 @@ def setup():
                 "Could not set locale to English/United States. "
                 "Some date-related tests may fail.")
 
-    mpl.use('Agg', warn=False)  # use Agg backend for these tests
+    mpl.use('Agg', force=True, warn=False)  # use Agg backend for these tests
```
Comment (efiring) [inline lib/matplotlib/testing/__init__.py]: The docstring for `mpl.use` says the use of `plt.switch_backend` is preferred over `force=True`; so why is the discouraged form being used in the tests?

File: lib/matplotlib/tests/test_text.py [hunk 3]
```diff
@@ -365,6 +365,10 @@ def test_annotation_negative_fig_coords():
 
 def test_text_stale():
     fig, (ax1, ax2) = plt.subplots(1, 2)
+    print(dict.__getitem__(plt.rcParams, "backend"))
+    print(plt.rcParams["backend"])
+    print(plt._backend_mod)
+    print(fig.canvas)
```
Comment (efiring) [inline lib/matplotlib/tests/test_text.py]: We don't usually include print calls in the tests, do we?  Are these left over from debugging, or are they intentional?  If the latter, given that they are unusual, do they merit a comment?

Comment (anntzer) [reply lib/matplotlib/tests/test_text.py]: that was for debugging the failure on travis, removing it.

File: lib/matplotlib/testing/__init__.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/testing/__init__.py]: The docstring should be updated (really there is no (actual) difference between use and switch_backend anymore after this PR (other than whether we silently or loudly fail or emit a warning) so I'd rather not point to switch_backend, but I know @tacaswell did add some pointers to it at some point so I'll let him decide.

File: lib/matplotlib/__init__.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/__init__.py]: @tacaswell you made this change I think?

File: lib/matplotlib/testing/__init__.py (hunk 2 above)
Comment (tacaswell) [reply lib/matplotlib/testing/__init__.py]: Removing that note from the docstring is something on my to-do list that I did not get to.

File: lib/matplotlib/__init__.py (hunk 1 above)
Comment (tacaswell) [reply lib/matplotlib/__init__.py]: I think this block can be dropped.
